### ISSUE TITLE
Feedback : test on LAN setup with 12 Wii + MKW

### ISSUE THREAD

**Issue by Dr-Wily** [barronwaffles/dwc_network_server_emulator#167]
I would post a feedback after a big test with 12 Wii with Mario Kart Wii. The purpose is to play with12 simultaneous players on a LAN without internet.

The setup :
-12 Wii
- 1 server, in two configurations 
  --> 1st : direct connection to the server ( DNS IP = server IP)
  --> 2nd : connection through a router
- on 12 profiles, 1 host all others friends and others only knows 1 same friend code. (the profile who hosts all friend code)
- test : worldwide game
- test : creating a game room to play GP with 12 players.
- the wifi AP is compose of 2 Netgear WNR1000v3. Each hosts 6 Wii.
- all Wii have the same configuration. And all profiles a totaly new
- The game run from an USB key with USB Loader GX (for using nossl patch)
- server setup : Debian 8 x64 with the latest version of the script. Server is a Intel C2D 3Ghz with 2Go of DDR2 and 100mb Ethernet.

Result :

Impossible to connect 12 Wii at the same time.

Profiles :
- All Wii can enter in WFC : no problem
- All profiles can see others online, but it has some delay (~10s)
- All profiles can create game room, and all other are able to see it.
- "Ramdomly", profile seem to be innacible with message code 61010. After this message the affected profile can not connect to WFC anymore. I must delete it and create another with a new friend code. Very annoying. A server reboot change nothing.
- the ban system uses the gateway IP to identify user. If all users have the same gateway, I can't ban one only. I ban all users who share the same gateway... stange choice.

Games :
- Worldwide game does not works. When all players chose "worldwide" at the same time, only 2 or 3 players are able to play together. I never able to start a worldwide game with 12 simultaneous players.
- hosted game works "half". Actually I able to connect only 4 Wii with "no" bug (except 61010) and a constant connection (no disconnect during gameplay). The max players I managed to enter in a game room is 9. After I get an error "impossible to join this player". When I start a game with 9 players, I have massive bugs :
- Ranking after race is totally crazy. Hosted profile on a Wii is always first All players end the race at first place and there is a disconnection. This is when the race can be ended.
- the most of time, race disconnects before the end.

Issues :
- error 61010 on profile, can't connect to WFC, I must delete and recreate profile
- can't connect 12 player in a game room and start a GP
- ramdom disconnections and erratic ranking in race above 6 players
- I get an error 84020 while searching for an opponent when 9 players are in a game room. If one player has this error then, all of the following players have the same error.

There is a best connection stability when all Wii are connected on the same Wifi AP. But 61010 and impossibility to start a race with 12 players still persist. Logs are not specially talkative about these issues.

have you some ideas on how to fix this and to play 12 players simultaneously ?

Were is stored players profile ?
How to restore a "defect" profile ? (61010) 

Thank !

**Comment by kyle95wm** [comment]
Nice test. This tells us a lot about the server's shortcomings. One thing I can comment on is the ban system. The reason it bans the gateway IP is because you're telling your DNS server to go out to the Internet and back in on the gateway therefore that's why you're banning your gateway address. To fix this tell your DNS server to redirect all queries for that Nintendo wifi address to your server's local address if that's how you're set up. If you for whatever reason are requiring the use of your external IP then I don't see an easy way to fix unless you set up a deprecate DNS server on your network and add the nintendosifi.net entry as mentioned above.

**Comment by kyle95wm** [comment]
Also 61010 for me always means Python crashed (an exception occurred) this be because I'm always messing with the code :p

If you don't know how to make your DNS server not use your external ip you have to edit /etc/dnsmasq.conf at the very bottom of the file change the IP and then write to the file and restart DNSMASQ

**Comment by kyle95wm** [comment]
tripple post - I am slowly working on a way to ban individual consoles as an alternative but things aren't going so well. The project is on https://github.com/kyle95wm/dwc_network_server_emulator/tree/wip but its hardly in a working fashion

**Comment by kyle95wm** [comment]
https://github.com/polaris-/dwc_network_server_emulator/pull/168 now we wait.....

**Comment by polaris-** [comment]
If you have any of the log files you could upload then I'd be interested in seeing them. Also, how did you connect the two routers exactly? Were they connected directly to each other somehow (cable?)?

**Comment by Dr-Wily** [comment]
kyle95wm --> OK, the ban system is minor for LAN play. It's just for information. For me, the most important thing is error 61010 and why I can't connect and race 12 players at the same time. Also, know where are stored players' profiles.

There is 2 majors codes 61010 and 84020 :
- 61010 prevents profile to connect to the server. The profile is definitively "banned". Once you encouter this code, it's impossible to reconnect. Even if the wii and/or the server is rebooted. I must delete it and create a new profile.
- 84020 : this code occur when I try to join more over 6 or 9 Wii in a game room. Sometime I can reach 9 Wii on a game room. After, systematically there is a 84020 code. This error spreads on all Wii. If I disconnect one player from the game room and try to reconnect it this error occurs.

84020 code disappear when I connect all Wii on the same router and the Wii get an IP from the same subnet as the router. But the problem persist and is substituted by another message "impossible to join this player".

Polaris -->

OK, where are stored the log (standard /var/log or specifics WFC log ?) and what are you needing ? My network setup :
- 1 level 2 switch. On it there are 4 devices conencted :
  ---> 2 routers 
  ---> the server
  ---> a laptop to read logs

The server hosts WFC and a DHCP. His IP is 192.168.0.154 and setup as static IP. Gateway and DNS on the server are setup to 192.168.10.154.

Firt router network configuration :
- "WAN" interface : get IP from DHCP (192.168.0.0/24)
- Wifi/LAN side : 192.168.1.0/24
- router act as DHCP on LAN side

Second router network configuration :
- "WAN" interface : get IP from DHCP (192.168.0.0/24)
- Wifi/LAN side : 192.168.2.0/24
- router act as DHCP on LAN side

The first time, I disabled the router DHCP for using the server's DCHP and Wii get IP directly from WFC server (192.168.10.154). Wii are on the same subnet that the server. A typical IP configuration on Wii is :

IP : 192.168.10.x/24
Gateway : 192.168.0.154
DNS1 : 192.168.10.154
DNS 2 : 192.168.10.154

But in this setup, there is a 84020 code where more than 6 players join a game room. (no problem to connect 12 player to WFC, it's just when I tryu to join game room.)

I can give you an acces to my server through SSH.

**Comment by polaris-** [comment]
@Dr-Wily There should be a folder called "logs" in the same folder where master_server.py, etc are located. It'll have a bunch of .log files inside of it. Please zip/rar/7z/whatever up the entire folder and send it over to me please.

**Comment by Dr-Wily** [comment]
OK.
- You can get fulls logs here : http://wellby.dyndns.org/dnl/logs-wfc-serv.7z
- A specific log when 61010 occurs : http://wellby.dyndns.org/dnl/log-wfc-61010.txt
- A specific log when the connection to WFC is OK : http://wellby.dyndns.org/dnl/log-wfc-no61010.txt
- A specific log when 84020 occurs : http://wellby.dyndns.org/dnl/log-wfc-84020.txt

You should look at dates between 21-05 and 24-05. Are the days where the server was used with 12 Wii (especially the 21 and 22). 23 and 24, server hosts only 4 Wii, but 61010 still persist randomly.

EDIT : I will make another big test on Friday / Saturday this week. If you change anything on your code and you would to test it, tell me so that I can update on my server.

**Comment by Dr-Wily** [comment]
OK, another feedback from a LAN party compose of 8 Wii.

Result --> success ! No disconnection, no corrupted profile.

The setup : 
- one wifi AP : a router/gateway between Wii network (wifi side) and server.
- 8 Wii (6 wireless, 2 wired)

On each Wii :
- I deleted profiles from previous LAN with 12 Wii
- I created news profiles
- With Ocarina I manage to unlock all characteres and vehicules before connect on the server on each savegame.
- then I connect each Wii on the server
- Finally, one profile hosts all friend codes, the others host only the friend code of the one who hosts all others.

All players can create a room, join it, start a game and play GP without any problem.

The only "glitch" I notice is at the name/mii associated with MKW profile. There is a remaining name/mii that I created on very firsts test of the server. This mii has been deleted from Wii, but it's still persists on the profile who host a game. Even if the mii associated with MKW profile has changed. Strange...
